### PR TITLE
[Feature]: Support MoCa

### PR DESCRIPTION
**Pull request by ZwwWayne**
This is the official code release of paper: [Multi-Modality Cut and Paste for 3D Object Detection](https://arxiv.org/abs/2012.12741). This PR includes the implementation of Hybrid Optimizer and multi-modality cut and paste on the KITTI dataset.

### CODE REVIEW

File: mmdet3d/core/optimizer/hybrid_constructor.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+from .hybrid_optimizer import HybridOptimizer
+
+
+@OPTIMIZER_BUILDERS.register_module()
+class HybridOptimizerConstructor(object):
+    """Special constructor for hybrid optimizers.
+    This constructor constructs hybrid optimizer for multi-modality
+    detectors. It builds separate optimizers for separate branchs for
+    different modalities. More details can be found in the ECCV submission
+    (to be release).
```
Comment (Tai-Wang) [inline mmdet3d/core/optimizer/hybrid_constructor.py:14]: Add a blank line before `Attributes`?

File: mmdet3d/core/optimizer/hybrid_constructor.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+          ),
+        )
+    """
+
+    def __init__(self, optimizer_cfg, paramwise_cfg=None):
+        if not isinstance(optimizer_cfg, dict):
+            raise TypeError('optimizer_cfg should be a dict',
+                            'but got {}'.format(type(optimizer_cfg)))
+        # assert paramwise_cfg is None, \
+        #     'Parameter wise config is not supported in Hybrid Optimizer'
```
Comment (Tai-Wang) [inline mmdet3d/core/optimizer/hybrid_constructor.py:66]: Remove the unused comment?

File: mmdet3d/core/optimizer/hybrid_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
+from mmcv.runner.optimizer import OPTIMIZERS
+from torch.optim import Optimizer
+
+
+@OPTIMIZERS.register_module()
+class HybridOptimizer(Optimizer):
+    """Hybrid Optimizer that contains multiple optimizers This optimizer
+    applies the hybrid optimzation for multi-modality models."""
```
Comment (Tai-Wang) [inline mmdet3d/core/optimizer/hybrid_optimizer.py:8]: Docstring for arguments or attributes

File: mmdet3d/datasets/pipelines/__init__.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 from mmdet.datasets.pipelines import Compose
 from .dbsampler import DataBaseSampler
-from .formating import Collect3D, DefaultFormatBundle, DefaultFormatBundle3D
+from .formating import Collect3D, DefaultFormatBundle3D
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/__init__.py:3]: This may influence nuImages models.

File: mmdet3d/datasets/pipelines/dbsampler.py [hunk 5]
```diff
@@ -314,3 +315,269 @@ def sample_class_v2(self, name, num, gt_bboxes):
             else:
                 valid_samples.append(sampled[i - num_gt])
         return valid_samples
+
+
+@OBJECTSAMPLERS.register_module()
+class MMDataBaseSampler(DataBaseSampler):
+
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/dbsampler.py:322]: Docstring

Comment (Tai-Wang) [reply mmdet3d/datasets/pipelines/dbsampler.py:322]: Also for the functions below.

File: mmdet3d/datasets/pipelines/formating.py [hunk 6]
```diff
@@ -6,78 +6,6 @@
 from mmdet.datasets.builder import PIPELINES
 from mmdet.datasets.pipelines import to_tensor
 
-PIPELINES._module_dict.pop('DefaultFormatBundle')
-
-
-@PIPELINES.register_module()
-class DefaultFormatBundle(object):
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/formating.py:13]: Does removing this class affect nuImages models?

File: mmdet3d/datasets/pipelines/loading.py [hunk 7]
```diff
@@ -18,45 +18,35 @@ class LoadMultiViewImageFromFiles(object):
...
+        if 'img_fields' not in results:
+            results['img_fields'] = []
+        for key in results['img_info']['filename']:
+            img = self._load_img(results['img_info']['filename'][key])
+            results['img_fields'].append(key)
+            results[key] = img
+            results['img_shape'] = img.shape
+            results['ori_shape'] = img.shape
+            # Set initial values for default meta_keys
+            results['pad_shape'] = img.shape
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/loading.py:49]: Other keys like `scale_factor` and `img_norm_cfg` are not used any longer?

File: mmdet3d/datasets/pipelines/test_time_aug.py [hunk 8]
```diff
@@ -77,11 +78,11 @@ def __call__(self, results):
         # modified from `flip_aug = [False, True] if self.flip else [False]`
         # to reduce unnecessary scenes when using double flip augmentation
         # during test time
-        flip_aug = [True] if self.flip else [False]
+        flip_aug = [False, True] if self.flip else [False]
```
Comment (Tai-Wang) [inline mmdet3d/datasets/pipelines/test_time_aug.py:81]: This seems to be a conflict. Please make sure this modification will not affect current models.

**Review by Tai-Wang — CHANGES_REQUESTED** [review]
Unit test may be added in next version.

File: mmdet3d/datasets/pipelines/__init__.py (hunk 4 above)
Comment (ZwwWayne) [reply mmdet3d/datasets/pipelines/__init__.py:3]: This will not as nuImages models should use a similar module as that in mmdet.

File: mmdet3d/models/fusion_layers/point_fusion.py [hunk 9]
```diff
@@ -28,9 +28,9 @@ def point_sample(
         img_features (torch.Tensor): 1 x C x H x W image features.
         points (torch.Tensor): Nx3 point cloud in LiDAR coordinates.
         lidar2img_rt (torch.Tensor): 4x4 transformation matrix.
-        img_scale_factor (torch.Tensor): Scale factor with shape of \
+        img_scale_factor (torch.Tensor): Scale factor with shape of
             (w_scale, h_scale).
-        img_crop_offset (torch.Tensor): Crop offset used to crop \
```
Comment (Tai-Wang) [inline mmdet3d/models/fusion_layers/point_fusion.py:33]: Should keep the '\' because it is needed when converting the API documentation?

File: configs/moca/dv_mvx-fpn_second_secfpn_hybrid_moca_2x8_80e_kitti-3d-3class.py [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+        dict(type='TensorboardLoggerHook')
+    ])
+# yapf:enable
+evaluation = dict(interval=1)
+# runtime settings
+total_epochs = 40
+dist_params = dict(backend='nccl')
+log_level = 'INFO'
+work_dir = None
+# You may need to download the model first is the network is unstable
```
Comment (Tai-Wang) [inline configs/moca/dv_mvx-fpn_second_secfpn_hybrid_moca_2x8_80e_kitti-3d-3class.py:162]: typo 'is the network is unstable'?

**Comment by filaPro** [comment]
Looks like `kitti_mm_dbinfos_train.pkl` is not the same with `kitti_dbinfos_train.pkl`. Are you going to provide this file or its generating script? Thanks.

**Comment by gujiaqivadin** [comment]
> Looks like `kitti_mm_dbinfos_train.pkl` is not the same with `kitti_dbinfos_train.pkl`. Are you going to provide this file or its generating script? Thanks.

It seems that moca needs instance_train.json which is the instance segmentation results of KITTI train split to produce this file. I also want to know whether moca would provide this file later.
@ZwwWayne Hello, wenwei. Do you have any plans for releasing moca? Or maybe providing a instance_train.json  or kitti_mm_dbinfos_train.pkl for pubilic to try some multi-modality 3d methods by our own. Thanks a lot!

**Comment by EchoQiHeng** [comment]
> > Looks like `kitti_mm_dbinfos_train.pkl` is not the same with `kitti_dbinfos_train.pkl`. Are you going to provide this file or its generating script? Thanks.
> 
> It seems that moca needs instance_train.json which is the instance segmentation results of KITTI train split to produce this file. I also want to know whether moca would provide this file later. @ZwwWayne Hello, wenwei. Do you have any plans for releasing moca? Or maybe providing a instance_train.json or kitti_mm_dbinfos_train.pkl for pubilic to try some multi-modality 3d methods by our own. Thanks a lot!

I according to this paper found the **instance-train. json** file of KINS dataset, then **Python tools/create_data.py --root-path --out_path --extra-tag kitti --version=mask.** Ultimately, **kitti_infos_xx.pkl** and **kitti_dbinfos_train.pkl** are generated, and **xx.bin.mask.png** and **xx.bin.png** are generated in **kitti_gt_dataset** file. But I don't know where the code is to generate **kitti_mm_dbinfos_train.pkl**.
If it's convenient for you, Could you please share the **kitti_mm_dbinfos_train.pkl** with me, or tell me the details of generating the **kitti_mm_dbinfos_train.pkl** from the **instance-train.json** file.
